bgpd/bgp_fsm.c

```diff
@@ -935,9 +935,25 @@ static void bgp_update_delay_process_status_change(struct peer *peer)
    read/write and timer thread. */
 void bgp_fsm_change_status(struct peer *peer, int status)
 {
+	struct bgp *bgp;
 
 	bgp_dump_state(peer, peer->status, status);
 
+	bgp = peer->bgp;
+	if (status == Established)
+		bgp->active_peers++;
+	else if ((peer->status == Established) && (status != Established))
+		bgp->active_peers--;
+
+	if (BGP_DEBUG(neighbor_events, NEIGHBOR_EVENTS))
+		zlog_debug("%s : vrf %u, active_peers %d", __func__,
+				bgp->vrf_id, bgp->active_peers);
+	/* If there are no active peers, set router ID to the value
+	 * provided by RIB
+	 */
+	if (bgp->active_peers == 0)
+		bgp_router_id_zebra_bump(bgp->vrf_id, NULL);
```

Comment:
nit: this is being called multiple times during the establishment of the first peer as it goes through the initialization FSM. Maybe we could try something like this:
```c
        bgp = peer->bgp;
        if (status == Established
            || ((peer->status == Established) && (status != Established))) {
                if (status == Established)
                        bgp->established_peers++;
                else
                        bgp->established_peers--;

                if (BGP_DEBUG(neighbor_events, NEIGHBOR_EVENTS))
                        zlog_debug("%s : vrf %u, established peers %u",
                                   __func__, bgp->vrf_id,
                                   bgp->established_peers);

                /*
                 * If there are no active peers, set router ID to the value
                 * provided by RIB.
                 */
                if (bgp->established_peers == 0)
                        bgp_router_id_zebra_bump(bgp->vrf_id, NULL);
        }
```

i.e. do nothing when the number of established peers didn't change.